src/meta/src/hummock/compaction/tier_compaction_picker.rs

```diff
@@ -12,202 +12,390 @@
+                continue;
+            }
 
-#[derive(Default)]
-pub struct TargetFilesInfo {
-    tables: Vec<SstableInfo>,
-    table_ids: HashSet<u64>,
-    compaction_bytes: u64,
-}
+            if l0.sub_levels[idx + 1].level_type == LevelType::Overlapping as i32 {
+                continue;
+            }
 
-impl TargetFilesInfo {
-    fn add_tables(&mut self, new_add_tables: Vec<SstableInfo>) {
-        for table in new_add_tables {
-            if self.table_ids.contains(&table.id) {
+            let min_overlap_picker = MinOverlappingPicker::new(
+                self.compact_task_id,
+                0,
+                0,
+                self.overlap_strategy.clone(),
+            );
+
+            let (select_tables, target_tables) = min_overlap_picker.pick_tables(
+                &l0.sub_levels[idx + 1].table_infos,
+                &level.table_infos,
+                level_handlers,
+            );
+
+            // only pick tables for trivial move
+            if select_tables.is_empty() || !target_tables.is_empty() {
                 continue;
             }
-            self.table_ids.insert(table.id);
-            self.compaction_bytes += table.file_size;
-            self.tables.push(table);
+
+            level_handlers[0].add_pending_task(self.compact_task_id, 0, &select_tables);
+            let input_levels = vec![
+                InputLevel {
+                    level_idx: 0,
+                    level_type: LevelType::Nonoverlapping as i32,
+                    table_infos: select_tables,
+                },
+                InputLevel {
+                    level_idx: 0,
+                    level_type: LevelType::Nonoverlapping as i32,
+                    table_infos: target_tables,
+                },
+            ];
+            return Some(CompactionInput {
+                input_levels,
+                target_level: 0,
+                target_sub_level_id: level.sub_level_id,
+            });
         }
+        None
     }
+}
+
+impl CompactionPicker for TierCompactionPicker {
+    fn pick_compaction(
+        &self,
+        levels: &Levels,
+        level_handlers: &mut [LevelHandler],
+    ) -> Option<CompactionInput> {
+        let l0 = levels.l0.as_ref().unwrap();
+        if l0.sub_levels.is_empty() {
+            return None;
+        }
+
+        if let Some(input) = self.pick_trivial_move_file(l0, level_handlers) {
+            return Some(input);
+        }
 
-    fn calc_inc_compaction_size(&self, new_add_tables: &[SstableInfo]) -> u64 {
-        new_add_tables
-            .iter()
-            .filter(|table| !self.table_ids.contains(&table.id))
-            .map(|table| table.file_size)
-            .sum()
+        if let Some(input) = self.pick_overlapping_level(l0, &mut level_handlers[0]) {
+            return Some(input);
+        }
+
+        self.pick_sharding_level(l0, &mut level_handlers[0])
     }
 }
 
+pub struct LevelCompactionPicker {
```

Comment:
Maybe we should move `LevelCompactionPicker` into a separate file.

Reply:
Good idea.

Reply:
done